index.js

```diff
@@ -1,227 +1,64 @@
-    }
-    return config;
   },
 
-  treeForAddon: function(tree) {
-    if (this._namespacingIsEnabled()) {
-      var allPodStyles = new Merge(this._allPodStyles, {
-        overwrite: true, // there are times (specifically with ember engines) where we run over the tree for twice. Should revist and find a way to prevent that in the future.
-        annotation: 'Merge (ember-component-css merge all process styles for a complete list of styles)'
-      });
-
-      var podNames = new ExtractNames(allPodStyles, {
-        classicStyleDir: this.classicStyleDir,
-        terseClassNames: this.terseClassNames,
-        annotation: 'Walk (ember-component-css extract class names from style paths)'
-      });
-
-      tree = new Merge([tree, podNames], {
-        overwrite: true,
-        annotation: 'Merge (ember-component-css merge names with addon tree)'
-      });
-    }
-    let checker = new VersionChecker(this);
-    let ember = checker.forEmber();
-    let superTree = this._super.treeForAddon.call(this, tree);
-    if (ember.isAbove('3.6.0')) {
-      return new Funnel(superTree, {
-        exclude: ['ember-component-css/initializers/route-styles.js'],
-        annotation:
-          "Funnel (ember-component-css exclude addon/initializers/route-styles.js in 3.6+)"
-      });
-    } else {
-      return new Funnel(superTree, {
-        exclude: ["ember-component-css/instance-initializers/route-styles.js"],
-        annotation:
-          "Funnel (ember-component-css exclude addon/instance-initializers/route-styles.js in < 3.6)"
-      });
-    }
+  _options(registry) {
+    return {
+      ...this._defaultOptions(registry),
+      ...this._overrideOptions(registry),
+    };
   },
 
-  treeForApp: function(tree) {
-    let checker = new VersionChecker(this);
-    let ember = checker.forEmber();
-    if (ember.isAbove('3.6.0')) {
-      return new Funnel(tree, {
-        exclude: ["initializers/route-styles.js"],
-        annotation:
-          "Funnel (ember-component-css exclude app/initializers/route-styles.js in 3.6+)"
-      });
-    } else {
-      return new Funnel(tree, {
-        exclude: ["instance-initializers/route-styles.js"],
-        annotation:
-          "Funnel (ember-component-css exclude app/instance-initializers/route-styles.js in < 3.6)"
-      });
-    }
+  isAddon() {
+    return Boolean(this.parent.parent);
   },
 
-  treeForParentAddonStyles: function(tree) {
-    let defaultTree = tree;
+  setupPreprocessorRegistry(type, registry) {
+    const options = this._options(registry);
 
-    if (this.parent.treeForAddonStyles) {
-      defaultTree = this.parent.treeForAddonStyles.apply(this.parent, arguments);
+    if (this.isAddon()) {
+      this.addAddonStyleHack(registry, options);
     }
+    
+    registry.add('css', new ColocateStyles(options));
+    registry.add('css', new NamespaceStyles(options));
+    registry.add('js', new ColocatedNamespaceObjects(options));
```

Comment:
```suggestion
    if(gte(3.13.0)) {
      registry.add('js', new ColocatedNamespaceObjects(options));
    }
```

Reply:
the namespace objects now are called their full style name with just a `.js` on the end, so there will be no conflict with those.